### ISSUE TITLE
Improve kwargs documentation

### ISSUE THREAD

**Issue by sergisiso** [stfc/PSyclone#2719]
We had the same problem on the PSyIR Nodes and Symbols kwargs, and the Transformations options which are passed to the super class. I don't like having them duplicated as we never check them when they are updated, or when new arguments added to the subclasses, and would be quite hard to review.

Adding a link to the super() is a good and simple solution (which some IDEs, static type tools may be able to follow). 

There is a more complex solution of programmatically modifying the `__doc__` to include the `BASE.__doc__` or even spit `__doc__` to separate the kwargs so that they can be included in the subclass docsting, but this may be over-engineering it, and probably not for this PR.

_Originally posted by @sergisiso in https://github.com/stfc/PSyclone/pull/2692#discussion_r1744024669_

**Comment by LonelyCat124** [comment]
There is also `inspect.getdoc` which might be better than just accessing __doc__. Form a cursory search:
> It cleans up the __doc__ by normalizing tabs to spaces and left shifting the doc body to remove common leading spaces.
which might be better if we just want to find arguments/types/raises and reformat them for this functions docs. The downside with all of this programatically is we aren't automating it. The best option we might have would be to create a `@inherit_parent_docs` or something to hide this in the actual Psyclone code, but I'm not sure how we'd do that.

**Comment by LonelyCat124** [comment]
I had a bit of a try for this - the wrapper seems very challenging to do because it loses self, which means the strategy I tried wasn't working.

From a general bit of code, you can find the parent class's docstring from the child function with:
`type(inspect.getmembers(y.apply)[22][1]).mro()[1].__dict__['apply'].__doc__`
where `inspect.getmembers(y.apply)[22]` is the `__self__` member, so `type(inspect.getmembers(y.apply)[22][1])` gives the object of `self` and `.mro()[1]` gets the parent class of `self`. `apply` here is specific function I'm looking at.

We could then pull the relevant bits from the inherited docs, but the thing I'm not clear how/where we would do that so that the documentation would see it.

**Comment by LonelyCat124** [comment]
Ok, a basic thing that just appends things is:

```
import re
docs = y.apply.__doc__
inspected_members = inspect.getmembers(y.apply)
parent_docstring = ""
for k, v in inspected_members:
    if k != "__self__":
        continue
    parent_docstring = type(v).mro()[1].__dict__[y.apply.__name__].__doc__
parent_lines = parent_docstring.splitlines()
for x, line in enumerate(parent_lines):
    if(":param" in line):
        param_str = re.search(":[a-zA-Z0-9\s]*:", line)
        if( param_str.group() in docs):
            continue
        docs += "\n" + line
        y = x+1
        while(y < len(parent_lines)):
            if(":param" in parent_lines[y] or ":raises" in parent_lines[y]):
                break
            docs += "\n" + parent_lines[y]
            y = y + 1
```

I think if we have things like `raises` or stuff after the type arguments in the `y.apply.__doc__` this won't order them correctly, but we could handle that later.

The issue I have is twofold:
1. I'm not sure if .__doc__. is writable - at least it doesn't seem to be on this object.
2. I still have no idea where it should be done.

**Comment by LonelyCat124** [comment]
A bit more cleaner, and lets me write over `__doc__` - working on the class level instead. We could use a class decorator for this potentially?
```
import re
docs = class2.apply.__doc__
parent_docstring = ""
parent_docstring = class2.__mro__[1].__dict__[y.apply.__name__].__doc__
parent_lines = parent_docstring.splitlines()
for x, line in enumerate(parent_lines):
    if(":param" in line):
        param_str = re.search(":[a-zA-Z0-9\s]*:", line)
        if( param_str.group() in docs):
            continue
        docs += "\n" + line
        z = x+1
        while(z < len(parent_lines)):
            if(":param" in parent_lines[z] or ":raises" in parent_lines[z]):
                break
            docs += "\n" + parent_lines[z]
            z = z + 1

#docs += "---\n" + parent_docstring
print(docs)

class2.apply.__doc__ = docs
print(y.apply.__doc__)
```

**Comment by sergisiso** [comment]
I was thinking something like:
```
class A:
    _trans_options = "\nOption1: Option1 doc"
    def apply():
        '''Apply Class A'''

A.apply.__doc__ += str(A._trans_options)

class B(A):
    _trans_options = A._trans_options + "\nOption2: Option2 doc"
    def apply():
        '''Apply Class B'''

B.apply.__doc__ += str(B._trans_options)
```

which produces:
```
pydoc test.B | cat
Help on class B in test:

test.B = class B(A)
 |  Method resolution order:
 |      B
 |      A
 |      builtins.object
 |
 |  Methods defined here:
 |
 |  apply()
 |      Apply Class B
 |      Option1: Option1 doc
 |      Option2: Option2 doc
```

**Comment by LonelyCat124** [comment]
Yeah, my only wondering was if class A and class B both defined `Option1`, and maybe had different defaults (which are usually in the docstring) or something, but whether thats a case we will ever hit or not I'm not sure.

It is a lot simpler of an implementation though.

**Comment by LonelyCat124** [comment]
We could do both options through a decorator on the class to automate/hide the detail. The one thing I'm unsure on so far is how inheritance works for a decorator

**Comment by LonelyCat124** [comment]
@sergisiso I had a quick look at this now and threw something together for general functions, but I'm not sure it would work for `__init__` kwargs (which I think is this use case?)

Since:
```
>>> print(Routine.__doc__)

    A sub-class of a Schedule that represents a subroutine, function or
    program unit.

    :param symbol: the Symbol used to represent this Routine in its Container.
    :type symbol: :py:class:`psyclone.psyir.symbols.RoutineSymbol`
    :param Optional[bool] is_program: whether this Routine represents the
                                      entry point into a program (e.g.
                                      Fortran Program or C main()). Default is
                                      False.
    :param kwargs: additional keyword arguments provided to the super class.
    :type kwargs: unwrapped dict.

    :raises TypeError: if any of the supplied arguments are of the wrong type.

>>> print(Routine.__init__.__doc__)
None
```

This is again a class-level inheritance deal, not just a general function one. Is this the only use-case from here?
